File: apps/android/app/src/main/java/dev/flextranslate/ui/i18n/AppLanguage.kt
package dev.flextranslate.ui.i18n

import android.content.Context
import java.util.Locale

/**
 * Язык, на котором рисуется весь интерфейс. Это выбор уровня приложения В РАНТАЙМЕ (тумблер RU/EN
 * внутри приложения), не зависящий ни от системной локали устройства, ни от языков перевода
 * (источник/цель — [dev.flextranslate.ui.FlexLanguage]). Переключение меняет [LocalStrings] в корне
 * приложения, и все экраны рекомпозятся на новом языке, не теряя состояния.
 */
enum class AppLanguage(val code: String, val nativeLabel: String) {
    RU("ru", "Русский"),
    EN("en", "English"),
    ;

    companion object {
        /**
         * Разумный дефолт для свежей установки: русский, если основная локаль устройства русская,
         * иначе английский. Применяется только пока пользователь сам не выбрал язык.
         */
        fun fromSystem(locale: Locale = Locale.getDefault()): AppLanguage =
            if (locale.language.equals(RU.code, ignoreCase = true)) RU else EN

        fun fromCode(code: String?): AppLanguage =
            entries.firstOrNull { it.code.equals(code, ignoreCase = true) } ?: fromSystem()
    }
}

/**
 * Хранит выбор [AppLanguage] между запусками через [android.content.SharedPreferences]. Пока
 * пользователь не выбрал явно, [load] откатывается к [AppLanguage.fromSystem]. Compose здесь не
 * нужен — живое состояние держит композиция, а это только читает/пишет постоянное значение.
 */
class AppLanguageStore(context: Context) {

    private val prefs = context.applicationContext
        .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    /** Сохранённый язык, либо дефолт из системной локали, если ещё ничего не выбрано. */
    fun load(): AppLanguage {
        val stored = prefs.getString(KEY_LANGUAGE, null) ?: return AppLanguage.fromSystem()
        return AppLanguage.fromCode(stored)
    }

    fun save(language: AppLanguage) {
        prefs.edit().putString(KEY_LANGUAGE, language.code).apply()
    }

    private companion object {
        const val PREFS_NAME = "flex_ui_prefs"
        const val KEY_LANGUAGE = "app_language"
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/i18n/AppLanguage.swift
import Foundation
import SwiftUI

/// Язык интерфейса всего приложения. Это выбор уровня приложения в рантайме (тумблер RU/EN
/// внутри приложения), не связанный ни с системной локалью устройства, ни с языками
/// перевода (FlexLanguage). Переключение меняет LocalStrings в корне приложения — все экраны
/// пересобираются на новом языке, не теряя состояния.
enum AppLanguage: String, CaseIterable, Sendable {
    case ru = "ru"
    case en = "en"

    var nativeLabel: String {
        switch self {
        case .ru: return "Русский"
        case .en: return "English"
        }
    }

    /// Разумный дефолт при чистой установке: русский, если основная локаль устройства русская,
    /// иначе английский. Нужен только пока пользователь сам язык не выбрал.
    static func fromSystem() -> AppLanguage {
        let code = Locale.current.language.languageCode?.identifier ?? ""
        return code.lowercased() == "ru" ? .ru : .en
    }

    static func fromCode(_ code: String?) -> AppLanguage {
        guard let code else { return fromSystem() }
        return AppLanguage(rawValue: code.lowercased()) ?? fromSystem()
    }
}

/// Хранит выбранный язык между запусками через UserDefaults.
/// Пока пользователь не выбрал явно, load() откатывается к AppLanguage.fromSystem().
/// На SwiftUI тут не завязано — живое состояние держит композиция, а это только
/// читает/пишет сохранённое значение.
final class AppLanguageStore: @unchecked Sendable {
    static let shared = AppLanguageStore()

    private let key = "flex_app_language"

    private init() {}

    func load() -> AppLanguage {
        let stored = UserDefaults.standard.string(forKey: key)
        return AppLanguage.fromCode(stored)
    }

    func save(_ language: AppLanguage) {
        UserDefaults.standard.set(language.rawValue, forKey: key)
    }
}
